src/plugin.rs

```diff
@@ -1,29 +1,37 @@
 use std::collections::BTreeSet;
-use std::sync::Arc;
 
-use async_trait::async_trait;
 use maplit::btreeset;
+use tokio::sync::mpsc;
+
+use crate::prelude::*;
 
-use crate::client::{ClientConfig, Context};
-use crate::error::Result;
 use crate::plugins;
 
 #[async_trait]
-pub trait Plugin: Sync + Send {
+pub trait Plugin {
     fn new_from_env() -> Result<Self>
     where
         Self: Sized;
 
-    async fn handle_connect(&self) -> Result<()> {
-        Ok(())
-    }
+    async fn run(self, stream: Receiver<Arc<Context>>) -> Result<()>;
+}
 
-    async fn handle_message(&self, _ctx: &Arc<Context>) -> Result<()> {
-        Ok(())
-    }
+type PluginMeta = (
+    tokio::sync::mpsc::Sender<Arc<Context>>,
+    tokio::task::JoinHandle<Result<()>>,
+);
```

Comment:
Would prefer this be a struct with named members. Shouldn't be any additional cost to creation.